tutorials/3-intro-to-brian-simulations.ipynb

```diff
@@ -0,0 +1,616 @@
+    "net.run(1*second)\n",
+    "# And keep a copy of those spikes\n",
+    "spikes_i = MP.i\n",
+    "spikes_t = MP.t\n",
+    "# Now construct the network that we run each time\n",
+    "# SpikeGeneratorGroup gets the spikes that we created before\n",
+    "SGG = SpikeGeneratorGroup(num_inputs, spikes_i, spikes_t)\n",
+    "eqs = '''\n",
+    "dv/dt = -v/tau : 1\n",
+    "'''\n",
+    "G = NeuronGroup(1, eqs, threshold='v>1', reset='v=0', method='linear')\n",
+    "S = Synapses(SGG, G, on_pre='v += weight')\n",
+    "S.connect()\n",
+    "M = SpikeMonitor(G)\n",
+    "# Store the current state of the network\n",
+    "net = Network(SGG, G, S, M)\n",
+    "net.store()\n",
+    "for tau in tau_range:\n",
+    "    # Restore the original state of the network\n",
+    "    net.restore()\n",
+    "    # Run it with the new value of tau\n",
+    "    net.run(1*second)\n",
+    "    output_rates.append(M.num_spikes/second)\n",
+    "plot(tau_range/ms, output_rates)\n",
+    "xlabel(r'$\\tau$ (ms)')\n",
+    "ylabel('Firing rate (sp/s)');"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "You can see that now there is much less noise and it increases monotonically because the input spikes are the same each time, meaning we're seeing the effect of the time constant, not the random spikes.\n",
+    "\n",
+    "Note that in the code above, we created `Network` objects. The reason is that in the loop, if we just called `run` it would try to simulate all the objects, including the Poisson neurons ``P``, and we only want to run that once. We use `Network` to specify explicitly which objects we want to include.\n",
+    "\n",
+    "The techniques we've looked at so far are the conceptually most simple way to do multiple runs, but not always the most efficient. Since there's only a single output neuron in the model above, we can simply duplicate that output neuron and make the time constant a parameter of the group."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "start_scope() \n",
+    "num_inputs = 100\n",
+    "input_rate = 10*Hz\n",
+    "weight = 0.1\n",
+    "tau_range = linspace(1, 10, 30)*ms\n",
+    "num_tau = len(tau_range)\n",
+    "P = PoissonGroup(num_inputs, rates=input_rate)\n",
+    "# We make tau a parameter of the group\n",
+    "eqs = '''\n",
+    "dv/dt = -v/tau : 1\n",
+    "tau : second\n",
+    "'''\n",
+    "# And we have num_tau output neurons, each with a different tau\n",
+    "G = NeuronGroup(num_tau, eqs, threshold='v>1', reset='v=0', method='euler')\n",
+    "G.tau = tau_range\n",
+    "S = Synapses(P, G, on_pre='v += weight')\n",
+    "S.connect()\n",
+    "M = SpikeMonitor(G)\n",
+    "# Now we can just run once with no loop\n",
+    "run(1*second)\n",
+    "# We extract the dictionary of spike trains to get the number\n",
+    "# of spikes for each output neuron\n",
+    "trains = M.spike_trains()\n",
```

Comment:
Not sure whether this is intentional to introduce the `spike_trains` method, but it would be easier to use `M.count`. Unfortunately, I just realized that the `count` attribute does not appear in the reference documentation for `SpikeMonitor` (or `EventMonitor`) because it is using our general `Variable` mechanism. We should look into adding it manually.